history.go

```diff
@@ -0,0 +1,75 @@
+package merkletree
+
+import (
+	"errors"
+
+	"github.com/coniks-sys/libmerkleprefixtree-go/crypto"
+)
+
+var (
+	ErrHistoryExisted = errors.New("[merkletree] trying to override existing history")
+	ErrBadEpoch       = errors.New("[merkletree] next epoch's STR has bad epoch ")
+)
+
+type History struct {
+	tree          *MerkleTree
+	currentSTR    *SignedTreeRoot
+	epochInterval int64
+}
+
+func NewHistory(m *MerkleTree, key crypto.SigningKey, startEp, epInterval int64) *History {
+	h := new(History)
+	h.tree = m
+	h.epochInterval = epInterval
+	h.currentSTR = newSTR(m, startEp, 0, make([]byte, crypto.HashSizeByte), key)
+	return h
+}
+
+func (h *History) UpdateHistory(m *MerkleTree, nextEp int64) error {
```

Comment:
> At t == 1, we take a snapshot (clone) of the tree as it is and sign the root.

So this STR is the STR for the tree at `t == 1`. The temporary bindings (namely `TB_0`) which is set in the tree at `t == 1` (namely `tree_1`) would belong to the tree at `t == 0` (namely `tree_0`), since we inserted them into the tree before cloning the tree. 

As a result, the `STR[0]` of `tree_0` (or `array[0]` as you called) is not computed based on `TB_0`, whereas the `STR[1]` of `tree_1` would be computed based on `TB_0`. 

Is that right?

Btw, thank you for your patience so far 😭